lib/Parser/parse.cpp

```diff
@@ -2076,6 +2117,554 @@ ParseNodePtr Parser::ParseMetaProperty(tokens metaParentKeyword, charcount_t ich
+{
+    bool parsedNamespaceOrNamedImport = false;
+
+    switch (m_token.tk)
+    {
+    case tkID:
+        // This is the default binding identifier.
+
+        // If we already saw a comma in the import clause, this is a syntax error.
+        if (parsingAfterComma)
+        {
+            Error(ERRsyntax);
+        }
+
+        if (buildAST)
+        {
+            IdentPtr localName = m_token.GetIdentifier(m_phtbl);
+            IdentPtr importName = wellKnownPropertyPids.default;
+            ParseNodePtr declNode = CreateModuleImportDeclNode(localName);
+
+            AddModuleImportEntry(importEntryList, importName, localName, nullptr, declNode);
+        }
+        
+        break;
+    
+    case tkLCurly:
+        // This begins a list of named imports.
+        ParseNamedImportOrExportClause<buildAST>(importEntryList, nullptr, false);
+
+        parsedNamespaceOrNamedImport = true;
+        break;
+
+    case tkStar:
+        // This begins a namespace import clause.
+        // "* as ImportedBinding"
+
+        // Token following * must be the identifier 'as'
+        m_pscan->Scan();
+        if (m_token.tk != tkID || wellKnownPropertyPids.as != m_token.GetIdentifier(m_phtbl))
+        {
+            Error(ERRsyntax);
+        }
+
+        // Token following 'as' must be a binding identifier.
+        m_pscan->Scan();
+        ChkCurTokNoScan(tkID, ERRsyntax);
+
+        if (buildAST)
+        {
+            IdentPtr localName = m_token.GetIdentifier(m_phtbl);
+            IdentPtr importName = CreatePid(L"*", 1);
+            ParseNodePtr declNode = CreateModuleImportDeclNode(localName);
+
+            AddModuleImportEntry(importEntryList, importName, localName, nullptr, declNode);
+        }
+
+        parsedNamespaceOrNamedImport = true;
+        break;
+
+    default:
+        Error(ERRsyntax);
+    }
+
+    m_pscan->Scan();
+
+    if (m_token.tk == tkComma)
+    {
+        // There cannot be more than one comma in a module import clause.
+        // There cannot be a namespace import or named imports list on the left of the comma in a module import clause.
+        if (parsingAfterComma || parsedNamespaceOrNamedImport)
+        {
+            Error(ERRsyntax);
+        }
+
+        m_pscan->Scan();
+
+        ParseImportClause<buildAST>(importEntryList, true);
+    }
+}
+
+bool Parser::IsImportOrExportStatementValidHere()
+{
+    // Import must be located in the global scope of the module.
+    return GetCurrentFunctionNode()->nop == knopProg
+        && (this->m_grfscr & fscrEvalCode) != fscrEvalCode
+        && this->m_tryCatchOrFinallyDepth == 0
+        && this->m_currentBlockInfo->pBlockInfoOuter == nullptr;
+}
+
+template<bool buildAST>
+ParseNodePtr Parser::ParseImportDeclaration()
+{
+    Assert(m_scriptContext->GetConfig()->IsES6ModuleEnabled());
+    Assert(m_token.tk == tkIMPORT);
+
+    if (!IsImportOrExportStatementValidHere())
```

Comment:
Supposedly it should always be true since we don't defer global-scope. If buildAST is false, this should return us the deferred function which isn't knopProg and which indicates the import is inside a function. I suppose we could check !buildAST here, as well.